### PR TITLE
SPHERE/IFS support 

### PR DESCRIPTION
**Pull request by arainot**
Hi Carlos,

Here I've added all the updated modules of VIP with IFS support. Let me know I have to make any modifications.

Alan

### CODE REVIEW

**Comment by carlgogo** [comment]
Hi @arainot, I made specific comments on different sections of the PR. My impression is that the PR is too big and complex, i.e. too many changes at once. I would propose to close this PR and do a few smaller PRs focused on specific functions (also making sure that only the necessary changes are done). It will be more tractable this way. Let me know if you agree. Cheers

**Comment by arainot** [comment]
Yes indeed that would be easier. I can't see your comments though, they don't appear on my PR.

**Comment by carlgogo** [comment]
@arainot indeed the comments were made but not submitted. You should be able to see them now. Then let's leave this PR pending here. Wait for the next release of VIP, it should be done later today. And start from there your focused PRs. Thanks!

**Comment by arainot** [comment]
Yes I see them now. Alright, sounds good!

File: vip/pca/utils_pca.py [hunk 1]
```diff
@@ -204,26 +230,26 @@ def matrix_scaling(matrix, scaling):
...
-    
-    return matrix 
+
+    return matrix
 
 
 def prepare_matrix(array, scaling=None, mask_center_px=None, mode='fullfr',
                    annulus_radius=None, annulus_width=None, verbose=True):
-    """ Builds the matrix for the SVD/PCA and other matrix decompositions, 
+    """ Builds the matrix for the SVD/PCA and other matrix decompositions,
```
[the remark this answers is not in the document]
Comment (arainot) [reply vip/pca/utils_pca.py]: It's for the contrast curves if I remember correctly

File: vip/pca/utils_pca.py [hunk 2]
```diff
@@ -320,96 +382,96 @@ def reconstruction(ncomp, U, S, V, var=1):
...
         M = np.dot(matrix, matrix.T)                             # covariance matrix
         e, EV = linalg.eigh(M)                                   # eigenvalues and eigenvectors
         pc = np.dot(EV.T, matrix)                                # PCs using a compact trick when cov is MM'
-        V = pc[::-1]                                             # reverse since last eigenvectors are the ones we want 
-        S = np.sqrt(e)[::-1]                                     # reverse since eigenvalues are in increasing order 
+        V = pc[::-1]                                             # reverse since last eigenvectors are the ones we want
+        S = np.sqrt(e)[::-1]                                     # reverse since eigenvalues are in increasing order
         if debug: reconstruction(ncomp, None, S, None)
-        for i in range(V.shape[1]):
+        for i in xrange(V.shape[1]):
```
[the remark this answers is not in the document]
Comment (arainot) [reply vip/pca/utils_pca.py]: That was your previous code, I thought git would have correct that. No worries, I'll do it by hand.

File: vip/phot/contrcurve.py [hunk 3]
```diff
@@ -332,31 +321,33 @@ def contrast_curve(cube, angle_list, psf_template, fwhm, pxscale, starphot,
             plt.ylabel('Delta magnitude')
             plt.gca().invert_yaxis()
             plt.grid('on', which='both', alpha=0.2, linestyle='solid')
-            ax3.set_xlim(0, np.max(rad_samp*pxscale))
+            ax3.set_xlim(0, np.max(rad_samp[:-30]*pxscale))
```
[the remark this answers is not in the document]
Comment (arainot) [reply vip/phot/contrcurve.py]: Sorry that was a personal test for plots, removed.

File: vip/preproc/recentering2.py [hunk 4]
```diff
@@ -1,153 +0,0 @@
-#! /usr/bin/env python
```
[the remark this answers is not in the document]
Comment (arainot) [reply vip/preproc/recentering2.py]: Oh yeah seems like it disappeared, weird

File: vip/phot/contrcurve.py [hunk 5]
```diff
@@ -332,31 +321,33 @@ def contrast_curve(cube, angle_list, psf_template, fwhm, pxscale, starphot,
...
+    datafr2 = DF(data=thruput_interp, columns=['throughput'])
+    datafr3 = DF(data=rad_samp, columns=['distance'])
+    datafr4 = DF(data=noise_samp_sm, columns=['noise'])
     if student:
-        datafr = pd.DataFrame({'sensitivity (Gauss)': cont_curve_samp,
-                               'sensitivity (Student)':cont_curve_samp_corr,
-                               'throughput': thruput_interp,
-                               'distance': rad_samp, 'noise': noise_samp_sm,
-                               'sigma corr':sigma_corr})
+        datafr1 = DF(data=cont_curve_samp_corr, columns=['sensitivity (Student)'])
```
[the remark this answers is not in the document]
Comment (arainot) [reply vip/phot/contrcurve.py]: Seems like git doesn't like this chunk, I'll have a look

File: vip/preproc/subsampling.py [hunk 6]
```diff
@@ -14,11 +14,11 @@
 import numpy as np
 
 
-def cube_collapse(cube, mode='median', n=50):
-    """ Collapses a cube into a frame (3d array to 2d array) depending on the 
-    parameter *mode*. It's possible to perform a trimmed mean combination of the 
+def cube_collapse(cube, mode='median', n=50, wl_cube=False):
```
[the remark this answers is not in the document]
Comment (arainot) [reply vip/preproc/subsampling.py]: I'll add the documentation. I needed to collapse my frames in order to get the best estimate for the photometry of my companions and PSF

**Comment by carlgogo** [comment]
Hi @arainot. I will close this PR. I suggest to start a new fork (on version 0.8.2) and send new PRs focused on single functions. Also, the functionalities need to satisfy general needs. To give you one example, you could start with sending a PR with the 4D injection of companions that we coded in Leuven a few months ago. It will be easier to review a single function and to merge your PR. Pay attention to the fact that the code must be py2 and py3 compatible.

**Pull request closed without merging**